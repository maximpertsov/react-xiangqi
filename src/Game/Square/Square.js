import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';

const SELECTION_GREEN = 'rgba(152, 251, 152, 0.3)';

const Wrapper = styled.div(
  {
    backgroundPosition: 'center',
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'contain',
    display: 'flex',
    justifyContent: 'center',
    padding: '0px;',
    margin: '0px;',
  },
  ({ selected, targeted }) => ({
    backgroundColor: (selected ? SELECTION_GREEN : 'none'),
    outline: (targeted ? `3px dotted ${SELECTION_GREEN}` : 'none'),
    outlineOffset: (targeted ? '-2px' : 'none'),
  }),
);

const Dot = styled.div`
  width:50%;
  height:50%;
  position:relative;
  top:50%;
  transform:translateY(-50%);
  border-radius:50%;
  background:${SELECTION_GREEN};
`;

class Square extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
    this.state = {
      selected: false,
    };
  }

  componentWillReceiveProps(nextProps) {
    const { slot, selectedSlot } = nextProps;

    if (this.isOccupied() && slot === selectedSlot) {
      this.setState((prevState) => ({ selected: !prevState.selected }));
    } else {
      this.setState({ selected: false });
    }
  }

  handleClick() {
    const { handleSquareClick } = this.props;
    handleSquareClick(this);
  }

  isOccupied() {
    const { piece } = this.props;
    return piece !== undefined;
  }

  isSelected() {
    const { selected } = this.state;
    return selected;
  }

  clickWillUnselect() {
    return this.isOccupied() && this.isSelected();
  }

  renderSquareElement() {
    const { slot, piece, targets } = this.props;
    if (this.isOccupied()) return piece;
    if (targets.includes(slot)) return (<Dot />);
    return (<div />);
  }

  render() {
    const { selected } = this.state;
    const { slot, targets } = this.props;
    const targeted = (this.isOccupied() && targets.includes(slot));

    return (
      <Wrapper
        className="Square"
        onClick={this.handleClick}
        selected={selected}
        targeted={targeted}
      >
        {this.renderSquareElement()}
      </Wrapper>
    );
  }
}

Square.propTypes = {
  handleSquareClick: PropTypes.func.isRequired,
  piece: PropTypes.element,
  slot: PropTypes.number.isRequired,
  selectedSlot: PropTypes.number,
  targets: PropTypes.arrayOf(PropTypes.number).isRequired,
};

Square.defaultProps = {
  piece: undefined,
  selectedSlot: null,
};

export default Square;
