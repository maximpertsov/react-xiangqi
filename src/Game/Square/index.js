/** @jsx jsx */
import { jsx } from '@emotion/core';

import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import XiangqiPiece from '../Piece/Piece';

const LAST_MOVE_COLOR = 'rgb(170,143,121,0.3)';
const SELECTION_GREEN = 'rgba(30,179,0,0.3)';
const IN_CHECK_RED = 'red';

const fillParentElement = {
  height: '100%',
  width: '100%',
  position: 'absolute',
};

const LastMoveIndicator = styled.div({
  backgroundColor: LAST_MOVE_COLOR,
  zIndex: '-1',
  ...fillParentElement,
});

const SelectionIndicator = styled.div({
  backgroundColor: SELECTION_GREEN,
  zIndex: '-1',
  ...fillParentElement,
});

const TargetIndicator = styled.div(({ occupied }) => ({
  ...(occupied) ? {
    outline: `2px dotted ${SELECTION_GREEN}`,
    ...fillParentElement,
  } : {
    backgroundColor: SELECTION_GREEN,
    borderRadius: '50%',
    height: '50%',
    width: '50%',
    position: 'relative',
    top: '50%',
    transform: 'translateY(-50%)',
  },
}));

const KingCheckedIndicator = styled.div({
  outline: `2px dotted ${IN_CHECK_RED}`,
  ...fillParentElement,
});


const Square = ({
  handleClick, pieceCode, selected, inCheck, inLastMove, targeted, moveX, moveY,
}) => {
  const occupied = pieceCode !== undefined;
  const moving = moveX !== 0 || moveY !== 0;

  const selectedMoveX = selected ? moveX : 0;
  const selectedMoveY = selected ? moveY : 0;

  return (
    <div
      className="Square"
      onClick={handleClick}
      css={{
        display: 'flex',
        justifyContent: 'center',
        margin: '0px;',
        padding: '0px;',
        position: 'relative',
      }}
    >
      {inLastMove && <LastMoveIndicator />}
      {selected && !(moving) && <SelectionIndicator />}
      {
        occupied &&
          <XiangqiPiece
            moveX={selectedMoveX}
            moveY={selectedMoveY}
            code={pieceCode}
          />
      }
      {targeted && !(moving) && <TargetIndicator occupied={occupied} />}
      {inCheck && !(moving) && <KingCheckedIndicator />}
    </div>
  );
};

Square.propTypes = {
  handleClick: PropTypes.func.isRequired,
  pieceCode: PropTypes.string,
  selected: PropTypes.bool.isRequired,
  inCheck: PropTypes.bool.isRequired,
  inLastMove: PropTypes.bool.isRequired,
  targeted: PropTypes.bool.isRequired,
  moveX: PropTypes.number,
  moveY: PropTypes.number,
};

Square.defaultProps = {
  pieceCode: undefined,
  moveX: 0,
  moveY: 0,
};

export default Square;
