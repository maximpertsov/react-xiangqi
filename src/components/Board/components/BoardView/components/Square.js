import React, { useContext } from 'react';
import { useDrop } from 'react-dnd';
import PropTypes from 'prop-types';

import { SquareContext } from 'contexts/SquareProvider';

import DropIndicator from './DropIndicator';
import KingInCheckIndicator from './KingInCheckIndicator';
import LastMoveIndicator from './LastMoveIndicator';
import Piece from './Piece';
import SelectionIndicator from './SelectionIndicator';
import SquareView from './SquareView';
import TargetIndicator from './TargetIndicator';

// TODO make handle square click an action?
const Square = ({ handleSquareClick }) => {
  const { isTargeted, square } = useContext(SquareContext);
  const [{ isOver }, drop] = useDrop({
    accept: 'PIECE',
    drop: () => {
      handleSquareClick(square)();
    },
    collect: monitor => ({
      isOver: !!monitor.isOver(),
      canDrop: !!monitor.canDrop(),
    }),
  });

  return (
    <SquareView handleClick={handleSquareClick(square)} ref={drop}>
      <Piece />
      <DropIndicator over={isOver} targeted={isTargeted} />
      <LastMoveIndicator />
      <KingInCheckIndicator />
      <SelectionIndicator />
      <TargetIndicator />
    </SquareView>
  );
};

Square.propTypes = {
  handleSquareClick: PropTypes.func.isRequired,
};

export default Square;
