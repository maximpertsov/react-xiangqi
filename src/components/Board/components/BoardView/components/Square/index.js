import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { useSelector } from 'react-redux';

import { getIsMoving } from 'reducers';
import { fillParentElement } from 'commonStyles';

import XiangqiPiece from './components/Piece';

const SELECTION_COLOR = 'rgba(30,179,0,0.3)';
const IN_CHECK_COLOR = 'red';

const SelectionIndicator = styled.div({
  backgroundColor: SELECTION_COLOR,
  zIndex: '-1',
  ...fillParentElement,
});

const TargetOccupiedIndicator = styled.div({
  backgroundColor: SELECTION_COLOR,
  borderRadius: '50%',
  ...fillParentElement,
});

const TargetEmptyIndicator = styled.div({
  backgroundColor: SELECTION_COLOR,
  borderRadius: '50%',
  height: '50%',
  position: 'relative',
  top: '50%',
  transform: 'translateY(-50%)',
  width: '50%',
});

const TargetIndicator = ({ occupied }) =>
  occupied ? <TargetOccupiedIndicator /> : <TargetEmptyIndicator />;

TargetIndicator.propTypes = {
  occupied: PropTypes.bool.isRequired,
};

const KingCheckedIndicator = styled.div({
  backgroundColor: IN_CHECK_COLOR,
  borderRadius: '50%',
  zIndex: '-1',
  ...fillParentElement,
});

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  margin: 0px;
  padding: 0px;
  position: relative;
`;

const Square = ({
  children,
  handleClick,
  inCheck,
  pieceCode,
  selected,
  targeted,
}) => {
  const [moveX, moveY] = useSelector(state => state.animationOffset);
  const isMoving = useSelector(state => getIsMoving(state));
  const occupied = pieceCode !== undefined;

  const selectedMoveX = selected ? moveX : 0;
  const selectedMoveY = selected ? moveY : 0;

  /* eslint-disable jsx-a11y/no-static-element-interactions */
  /* eslint-disable jsx-a11y/click-events-have-key-events */
  return (
    <Wrapper className="Square" onClick={handleClick}>
      {children}
      {selected && !isMoving && <SelectionIndicator />}
      {occupied && (
        <XiangqiPiece
          moveX={selectedMoveX}
          moveY={selectedMoveY}
          code={pieceCode}
        />
      )}
      {targeted && !isMoving && <TargetIndicator occupied={occupied} />}
      {inCheck && !isMoving && <KingCheckedIndicator />}
    </Wrapper>
  );
  /* eslint-enable jsx-a11y/no-static-element-interactions */
  /* eslint-enable jsx-a11y/click-events-have-key-events */
};

Square.propTypes = {
  children: PropTypes.node.isRequired,
  handleClick: PropTypes.func.isRequired,
  pieceCode: PropTypes.string,
  selected: PropTypes.bool.isRequired,
  inCheck: PropTypes.bool.isRequired,
  targeted: PropTypes.bool.isRequired,
};

Square.defaultProps = {
  pieceCode: undefined,
};

export default Square;
