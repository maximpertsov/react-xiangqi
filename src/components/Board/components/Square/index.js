/** @jsx jsx */
import { jsx } from '@emotion/core';

import PropTypes from 'prop-types';
import styled from '@emotion/styled';

import XiangqiPiece from './components/Piece';

const LAST_MOVE_COLOR = 'rgba(201,255,229,1.0)';
const SELECTION_COLOR = 'rgba(30,179,0,0.3)';
const IN_CHECK_COLOR = 'red';

const fillParentElement = {
  height: '100%',
  width: '100%',
  position: 'absolute',
};

const LastMoveIndicator = styled.div({
  backgroundColor: LAST_MOVE_COLOR,
  borderRadius: '50%',
  zIndex: '-1',
  ...fillParentElement,
});

const SelectionIndicator = styled.div({
  backgroundColor: SELECTION_COLOR,
  zIndex: '-1',
  ...fillParentElement,
});

const TargetIndicator = styled.div(({ occupied }) => ({
  ...(occupied) ? {
    outline: `2px dotted ${SELECTION_COLOR}`,
    ...fillParentElement,
  } : {
    backgroundColor: SELECTION_COLOR,
    borderRadius: '50%',
    height: '50%',
    width: '50%',
    position: 'relative',
    top: '50%',
    transform: 'translateY(-50%)',
  },
}));

const KingCheckedIndicator = styled.div({
  outline: `2px dotted ${IN_CHECK_COLOR}`,
  ...fillParentElement,
});


const Square = ({
  handleClick, pieceCode, selected, inCheck, inLastMove, targeted, moveX, moveY,
}) => {
  const occupied = pieceCode !== undefined;
  const moving = moveX !== 0 || moveY !== 0;

  const selectedMoveX = selected ? moveX : 0;
  const selectedMoveY = selected ? moveY : 0;

  return (
    <div
      className="Square"
      onClick={handleClick}
      css={{
        display: 'flex',
        justifyContent: 'center',
        margin: '0px;',
        padding: '0px;',
        position: 'relative',
      }}
    >
      {inLastMove && <LastMoveIndicator />}
      {selected && !(moving) && <SelectionIndicator />}
      {
        occupied &&
          <XiangqiPiece
            moveX={selectedMoveX}
            moveY={selectedMoveY}
            code={pieceCode}
          />
      }
      {targeted && !(moving) && <TargetIndicator occupied={occupied} />}
      {inCheck && !(moving) && <KingCheckedIndicator />}
    </div>
  );
};

Square.propTypes = {
  handleClick: PropTypes.func.isRequired,
  pieceCode: PropTypes.string,
  selected: PropTypes.bool.isRequired,
  inCheck: PropTypes.bool.isRequired,
  inLastMove: PropTypes.bool.isRequired,
  targeted: PropTypes.bool.isRequired,
  moveX: PropTypes.number,
  moveY: PropTypes.number,
};

Square.defaultProps = {
  pieceCode: undefined,
  moveX: 0,
  moveY: 0,
};

export default Square;
