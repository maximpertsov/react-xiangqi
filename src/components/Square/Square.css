.Square {
  background: url("https://upload.wikimedia.org/wikipedia/commons/6/66/Xiangqi-_md.PNG");
  background-size: contain;
}
