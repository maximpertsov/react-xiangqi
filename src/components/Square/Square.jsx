import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';

const Wrapper = styled.div(
  {
    backgroundPosition: 'center',
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'contain',
    display: 'flex',
    justifyContent: 'center',
  },
  ({ selected }) => ({
    backgroundColor: (selected ? 'rgba(152, 251, 152, 0.3)' : 'none'),
  }),
);

class Square extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    const { piece, cellName, handleClick } = this.props;
    if (piece !== undefined) {
      handleClick(cellName);
    } else {
      handleClick(null);
    }
  }

  render() {
    const { piece, selected } = this.props;
    return (
      <Wrapper
        className="Square"
        onClick={this.handleClick}
        selected={selected}
      >
        {piece}
      </Wrapper>
    );
  }
}

Square.propTypes = {
  cellName: PropTypes.string.isRequired,
  handleClick: PropTypes.func.isRequired,
  piece: PropTypes.element,
  selected: PropTypes.bool.isRequired,
};

Square.defaultProps = {
  piece: undefined,
};

export default Square;
