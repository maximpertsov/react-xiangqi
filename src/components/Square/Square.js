import React, { Component } from 'react';
import './Square.css';

class Square extends Component {
  render() {
    return (
      <div className="Square">
        <img className="Piece" src={this.props.pieceImgSrc} />
      </div>
    );
  }
}

export default Square;
