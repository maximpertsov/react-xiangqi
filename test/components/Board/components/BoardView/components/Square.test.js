import React from 'react';
import { render } from 'enzyme';

import Square from 'components/Board/components/BoardView/components/Square';

describe('Square', () => {
  test('renders without crashing', () => {
    const wrapper = render(<Square />);
    expect(wrapper).toMatchSnapshot();
  });
});
