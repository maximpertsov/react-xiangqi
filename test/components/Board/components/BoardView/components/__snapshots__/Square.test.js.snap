// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Square renders without crashing 1`] = `
<div
  class="Square css-8516zs"
/>
`;
